### PR TITLE
Add inner modules

### LINKED ISSUE #74: Add inner modules
Add modules inside a single file:

```wgsl
#mod main
// ~.main module starts here
const A = 0;

#mod inner
// ~.inner module starts here
const B = 0;

#shader<compute> my_shader
// also possible to create multiple shaders in the same file
#import ~ // to import root module 
```

Duplicated module names should be detected.

### PR DESCRIPTION
**Pull request by Nicolas-Ferre**
Closes #74

### CODE REVIEW

File: crates/wgso/src/program/file.rs [hunk 1]
```diff
@@ -104,16 +75,44 @@ pub(crate) struct File {
...
+                errors.push(Error::DirectiveParsing(ParsingError {
+                    path: path.into(),
+                    span: offset..offset,
+                    message: "file should start with `#mod` or `#shader` directive".into(),
+                }));
+                return;
+            }
+            offset += line.len() + 1;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Nicolas-Ferre) [reply crates/wgso/src/program/file.rs:117]: Offset is not used outside the loop, so we should be good for the last line.
Unicode indexing will be applied in next commit

[the remark this answers is not in the document]
Comment (Nicolas-Ferre) [reply crates/wgso/src/program/file.rs:117]: Actually for now I will not fix for unicode indexes, as this is an edge case that shouldn't happen often

File: crates/wgso/tests/cases_valid/storages/main.wgsl [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+#mod array
+
+var<storage, read_write> primitive_array_value: array<u32, 10>;
+var<storage, read_write> struct_array_value: array<TestStruct, 10>;
+var<storage, read_write> unsized_array_value: array<TestStruct>;
+
+struct TestStruct {
+    f32_field: f32,
+}
```
[the remark this answers is not in the document]
Comment (Nicolas-Ferre) [reply crates/wgso/tests/cases_valid/storages/main.wgsl:17]: This is fine. WGSO introduces #mod and #shader annotations to define multiple independent modules inside a single WGSL file

File: crates/wgso/src/program/section.rs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                        Arc::new(section),
+                    )
+                })
+                .collect(),
+        }
+    }
+
+    pub(crate) fn get(&self, path: &(PathBuf, String)) -> &Arc<Section> {
+        &self.sections[path]
+    }
```
[the remark this answers is not in the document]
Comment (Nicolas-Ferre) [reply crates/wgso/src/program/section.rs:35]: This is internal code, so it should be fine

File: crates/wgso/src/program/wgsl.rs [hunk 4]
```diff
@@ -34,54 +34,51 @@ impl WgslModule {
...
     pub(crate) fn configure_buffer_types(&mut self) {
-        for file in &self.files {
-            let location_offset = Self::configure_buffer_type(&mut self.module, file, true, 0);
-            Self::configure_buffer_type(&mut self.module, file, false, location_offset);
+        let directive = &self.sections[0].directive;
+        if directive.kind() == DirectiveKind::RenderShader {
+            let location_offset = Self::configure_buffer_type(&mut self.module, directive, true, 0);
+            Self::configure_buffer_type(&mut self.module, directive, false, location_offset);
         }
     }
```
[the remark this answers is not in the document]
Comment (Nicolas-Ferre) [reply crates/wgso/src/program/wgsl.rs:42]: Shouldn't panic in practice

File: crates/wgso/src/error.rs [hunk 5]
```diff
@@ -178,8 +194,9 @@ impl Error {
         format!(
             "{}",
             Renderer::styled().render(message.footer(Level::Info.title(&format!(
-                "The error comes from file '{}'",
-                files[0].path.display()
+                "The error comes from `{}` module in file '{}'",
+                sections[0].directive.section_name().slice,
+                sections[0].path().display(),
             ))))
```
[the remark this answers is not in the document]
Comment (Nicolas-Ferre) [reply crates/wgso/src/error.rs:200]: Shouldn't panic in practice
